discovery/discovery.go

```diff
@@ -1,19 +1,49 @@
 package discovery
 
-// DiscoveryService is an interface
+import log "github.com/Sirupsen/logrus"
+
+// Service is an interface
 // which all service discovery backends must implement
-type DiscoveryService interface {
+type Service interface {
 	SendHeartbeat(service *ServiceDefinition)
 	CheckForUpstreamChanges(backendName string, backendTag string) bool
 	MarkForMaintenance(service *ServiceDefinition)
 	Deregister(service *ServiceDefinition)
 }
 
+// ServiceDefinition is the concrete service structure that is
+// registered with the service discovery backend.
 type ServiceDefinition struct {
 	ID        string
 	Name      string
 	Port      int
 	TTL       int
 	Tags      []string
-	IpAddress string
+	IPAddress string
+}
+
+// ServiceDiscoveryHandler parses a raw service discovery config
+type ServiceDiscoveryHandler func(interface{}) (Service, error)
+
+var backends = []string{}
+var discoveryHandlers = map[string]ServiceDiscoveryHandler{}
+
+// RegisterHandler registers a service discovery handler for a config key
+func RegisterHandler(name string, handler ServiceDiscoveryHandler) {
```

Comment:
This is going to sound nitpicky, but is "handler" really the terminology we want to use here? A handler sounds like it's something we're going to be dealing with on an ongoing basis at runtime (like the poll handlers). I'll admit I'm having trouble coming up with a better name (`Configurator` is a little silly. 😀  )